# Group 10 - First Year Project
### Data Science at ITU
## Project 1 - Road collisions analysis 2021 

The first project of Data Science's second semester required us to inform various city leaders around the United Kingdom about traffic fatalities and injuries; and give some insight for urban transport planning. This jupyter notebook presents Group 10's exploration of 2019 UK road collision data with complimentary analysis, visualisation and findings with relation to the city of Sheffield.


# Task 0 - Data filtering and cleaning
## Loading data
Our first task invovles loading our 3 csv files (downloaded from https://data.gov.uk/dataset/road-accidents-safety-data on the 4th January 2021) into a traversable data structure we can start working with; such as a dictionary. We will then need to start the process of filtering and cleaning our data.

In [29]:
import numpy as np
import pandas as pd

In [30]:
raw_data = {}

TITLES = ['accidents', 'casualties', 'vehicles'] # Capitalisation is the standard for constants

Pandas has a convinient method for reading in csv files and storing the data in a DataFrame structure offers flexibility.

In [31]:
raw_data['accidents'] = pd.read_csv("../data/raw/Road Safety Data - Accidents 2019.csv")
raw_data['casualties'] = pd.read_csv("../data/raw/Road Safety Data - Casualties 2019.csv")
raw_data['vehicles'] = pd.read_csv("../data/raw/Road Safety Data- Vehicles 2019.csv") # File name missing a space before hyphen 

In [32]:
for TITLE in TITLES:
    print(TITLE + "\tDataFrame shape: " + str(raw_data[TITLE].shape))

accidents	DataFrame shape: (117536, 32)
casualties	DataFrame shape: (153158, 16)
vehicles	DataFrame shape: (216381, 23)


 ### Data type insight?

In [33]:
#todo

## Locating our city
With our data loaded into a dictionary of DataFrames, we needed to locate the accidents that occured in Sheffield. To do this we made a mask identifying which accidents occured in the local district authority 215.


In [34]:
sheffield_mask = (raw_data['accidents']["Local_Authority_(District)"] == 215)

In [35]:
sheffield_data = {}

sheffield_data['accidents'] = raw_data['accidents'][sheffield_mask]
sheffield_data['accidents'][:5]

Accident_Index  Location_Easting_OSGR  Location_Northing_OSGR  \
44673  2019140805407               429140.0                381931.0   
44674  2019140805409               434290.0                385121.0   
44676  2019140805414               435393.0                391745.0   
44677  2019140805567               433949.0                396778.0   
44680  2019140805955               433662.0                389819.0   

       Longitude   Latitude  Police_Force  Accident_Severity  \
44673  -1.563884  53.333509            14                  2   
44674  -1.486208  53.361875            14                  3   
44676  -1.468895  53.421340            14                  1   
44677  -1.490081  53.466672            14                  2   
44680  -1.495146  53.404142            14                  3   

       Number_of_Vehicles  Number_of_Casualties        Date  ...  \
44673                   2                     1  01/01/2019  ...   
44674                   2                     1  01/01/2019  ...   
44676                   1                     1  01/01/2019  ...   
44677                   2                     1  01/01/2019  ...   
44680                   2                     1  03/01/2019  ...   

       Pedestrian_Crossing-Human_Control  \
44673                                  0   
44674                                  0   
44676                                  0   
44677                                  0   
44680                                  0   

      Pedestrian_Crossing-Physical_Facilities  Light_Conditions  \
44673                                       0                 1   
44674                                       0                 4   
44676                                       1                 4   
44677                                       0                 1   
44680                                       0                 1   

      Weather_Conditions  Road_Surface_Conditions  Special_Conditions_at_Site  \
44673                  1                        2                           0   
44674                  1                        1                           0   
44676                  1                        1                           0   
44677                  1                        1                           0   
44680                  1                        1                           0   

       Carriageway_Hazards  Urban_or_Rural_Area  \
44673                    0                    2   
44674                    0                    1   
44676                    0                    1   
44677                    0                    1   
44680                    0                    1   

       Did_Police_Officer_Attend_Scene_of_Accident  LSOA_of_Accident_Location  
44673                                            1                  E01007926  
44674                                            1                  E01008052  
44676                                            1                  E01008123  
44677                                            1                  E01007901  
44680                                            2                  E01008093  

[5 rows x 32 columns]

In [36]:
sheffield_data['accidents'].shape

(909, 32)

We can now see that there are 909 records of road traffic accidents occuring in Sheffield over 2019. The next step was using the primary key (accident_index) of the accidents to find the accidents in Sheffield that each casualty and vehicle linked to, in order to complete our sheffield_data dictionary.

In [37]:
sheffield_data['accidents']["Accident_Index"][:5]

44673    2019140805407
44674    2019140805409
44676    2019140805414
44677    2019140805567
44680    2019140805955
Name: Accident_Index, dtype: object

In [38]:
for TITLE in TITLES[1:]: sheffield_data[TITLE] = raw_data[TITLE][raw_data[TITLE]['Accident_Index'].isin(sheffield_data["accidents"]["Accident_Index"])]

In [39]:
for TITLE in TITLES: print("Sheffield " + TITLE + "\tDataFrame shape: " + str(sheffield_data[TITLE].shape))

Sheffield accidents	DataFrame shape: (909, 32)
Sheffield casualties	DataFrame shape: (1191, 16)
Sheffield vehicles	DataFrame shape: (1648, 23)


## Relational data table model 
In terms of the structure of the three databases or data tables, the accidents table's primary key is used as a foreign key by both the casualties and vehicles data tables, creating a one-to-many link between the accidents and both the casualties and vehicles tables. From this data table structure, we can infer that multiple casualties and vehicles can be involved in a single accident; which makes sense logically.

## Sanity check
Performing a sanity check on our Sheffield road accident data is a critical first step before we can begin to think about further analysing or drawing conclusions from the data. We may also have to sanity check all the data if we wish to draw conclusions therefrom.

In [40]:
#Check for empty strings in categorical fields  
#Check for NULL values                          #DONE 
#Check duplicate records                        #DONE
#Check non-linked records in secondary tables   

### Checking for null values in our road accident dataset

In [41]:
flag = False
for TITLE in TITLES:
    if raw_data[TITLE].isnull().values.any():
        for column in list(raw_data[TITLE].columns):
            nulls = sum(raw_data[TITLE][column].isnull())
            if sum(raw_data[TITLE][column].isnull()):
                print(TITLE + " - " + column + "\t has " + str(nulls) + " null values")
                flag = True
if not flag: print("There are no null values in the dataset")

accidents - Location_Easting_OSGR	 has 28 null values
accidents - Location_Northing_OSGR	 has 28 null values
accidents - Longitude	 has 28 null values
accidents - Latitude	 has 28 null values
accidents - Time	 has 63 null values
accidents - LSOA_of_Accident_Location	 has 5714 null values


### Checking for null values in our sheffield dataset

In [42]:
flag = False
for TITLE in TITLES:
    if sheffield_data[TITLE].isnull().values.any():
        for column in list(sheffield_data[TITLE].columns):
            nulls = sum(sheffield_data[TITLE][column].isnull())
            if sum(sheffield_data[TITLE][column].isnull()):
                print(TITLE + " - " + column + "\t has " + str(nulls) + " null values")
                flag = True
if not flag: print("There are no null values in the dataset")

There are no null values in the dataset


In [43]:
if len(sheffield_data["accidents"]["Accident_Index"]) == len((sheffield_data["accidents"]["Accident_Index"].unique())):
    print("No repeated indexes")
else: print("Repeated indexes detected")

No repeated indexes


## Numerical summary

In [44]:
import matplotlib.pyplot as plt

In [45]:
#Five Number Summary

In [46]:
#Box Plots

In [47]:
#Frequency Histogram

# Task 1 - Single variable analysis
### Report the frequency of road collisions in your city for 

1. different age groups 
2. in different times (during the day, the week, or year)
3. and for differences in one other condition.


# Task 2 - Associations
### Research whether there is a significant statistical association in your city, either
1. between some vehicle attributes and accident circumstances, or 
2. between some casualty attributes and accident circumstances. 

Report whether there is a statistically significant association between such variables or not, together with the appropriate statistical metric(s).
 
Discuss why this association, or the lack of
this association, is relevant for urban/transport planning.

# Task 3 - Map visualization
Visualize the reported collisions on a map of your city. Make a visual distinction between different
classes of casualty severity

# Task 4: Open question
Use the data to formulate, motivate, answer, and discuss another research question of your choice.
For example, compare your city to the whole UK, or investigate different collision participants (bicycle
vs car, bicycle-motored two-wheeler, pedestrian vs car, car vs car,…), or compare the age distribution
of casualties to the existing age distribution, or rank/compare how problematic different driver
demographics are (young/old male/female),…